README.md
# blackjack-clj

A clojure cli blackjack game

## Useage

#### Native:
Have [lein](https://leiningen.org/) installed.

    lein run

#### Docker:

    docker-compose build
    docker-compose run blackjack lein run

## Testing:
Have [lein](https://leiningen.org/) installed.

    lein run

## Caveats:
 - This game doesn't handle splits
 - This game dosen't handle the case where the dealer has blackjack properly:
   - If the player doesn't also have blackjack, they are allowed to play a hand and attempt to match the dealer's 21
   - If the player does have blackjack, the player wins


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2018-09-24
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2018-09-24
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/blackjack-clj/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/blackjack-clj/compare/0.1.0...0.1.1


test/blackjack_clj/core_test.clj
(ns blackjack-clj.core-test
  (:require [clojure.test :refer :all]
            [blackjack-clj.core :refer :all]))


test/blackjack_clj/cards_test.clj
(ns blackjack-clj.cards-test
  (:require [clojure.test :refer :all]
            [blackjack-clj.cards :refer :all]))

(def ace-hearts (list 0 hearts))
(def two-clubs (list 1 clubs))
(def king-spades (list 12 spades))
(def queen-hearts (list 11 hearts))
(def jack-clubs (list 10 clubs))
(def nine-diamonds (list 8 diamonds))

(deftest suit-test
  (testing "Suits are correctly returned by the suit function"
    (is (= hearts (suit ace-hearts)))
    (is (= clubs (suit two-clubs)))
    (is (= spades (suit king-spades)))))

(deftest display-value-test
  (testing "Names of card values are correct"
    (is (= "King" (display-value king-spades)))
    (is (= "Queen" (display-value queen-hearts)))
    (is (= "Jack" (display-value jack-clubs)))
    (is (= "Ace" (display-value ace-hearts)))
    (is (= "2" (display-value two-clubs)))))

(deftest display-card-test
  (testing "Human-readable card names are correct"
    (is (= "Ace of Hearts" (display-card ace-hearts)))
    (is (= "2 of Clubs" (display-card two-clubs)))
    (is (= "Jack of Clubs" (display-card jack-clubs)))
    (is (= "King of Spades" (display-card king-spades)))
    (is (= "9 of Diamonds" (display-card nine-diamonds)))))

(deftest hand-points-test
  (testing "Hands have their points counted correctly"
    (is (= 13 (hand-points (list ace-hearts two-clubs))))
    (is (= 12 (hand-points (list jack-clubs two-clubs))))
    (is (= 30 (hand-points (list king-spades jack-clubs queen-hearts))))
    (is (= 21 (hand-points (list jack-clubs ace-hearts))))
    (is (= 13 (hand-points (list ace-hearts two-clubs jack-clubs))))
    (is (= 13 (hand-points (list ace-hearts ace-hearts ace-hearts jack-clubs))))
    (is (= 21 (hand-points (list nine-diamonds
                                 nine-diamonds
                                 ace-hearts
                                 two-clubs))))))

(deftest hand-soft?-test
  (testing "Correctly reports hands are soft"
    (is (hand-soft? (list ace-hearts two-clubs)))
    (is (hand-soft? (list ace-hearts ace-hearts)))
    (is (hand-soft? (list jack-clubs ace-hearts))))
  (testing "Correctly reports hands are hard"
    (is (not (hand-soft? (list ace-hearts ace-hearts jack-clubs))))
    (is (not (hand-soft? (list jack-clubs nine-diamonds))))
    (is (not (hand-soft? (list two-clubs))))))


src/blackjack_clj/core.clj
(ns blackjack-clj.core
  (:require [clojure.string :refer [lower-case join]]
            [blackjack-clj.cards :as cards]))

(defn y-n
  "Prompts the user for input with $messages applied over str, followed by [Y/n]"
  [& messages]
  (println (apply str messages) "[Y/n]")
  (case (lower-case (read-line))
    "y" true
    "" true
    "n" false
    ;; Default
    (do (println "Sorry, I didn't understand that.")
        (apply y-n messages))))

(defn deal-card
  "Deals $player a card from $deck on to $hand
  returns '(hand deck)"
  ([player deck hand]
   (println player "delt:" (cards/display-card (first deck)))
   (deal-card deck hand))
  ([deck hand]
   (list (cons (first deck) hand) (rest deck))))

(defn deal
  "Returns (deck player-hand dealer-hand)"
  []
  (let [deck (cards/shuffled-deck)
        [player-hand deck] (deal-card "You're" deck (list))
        [dealer-hand deck] (deal-card "The dealer is" deck (list))
        [player-hand deck] (deal-card "You're" deck player-hand)
        [dealer-hand deck] (do (println "The dealer is delt a card face down.")
                               (deal-card deck dealer-hand))]
    (list deck player-hand dealer-hand)))

(defn play-card
  "Plays a player hand. Returns '(balance deck player-points bet dealer?)
  If bust, bet will be zero
  dealer? represents whether the dealer plays"
  [hand deck first? balance bet]
  (let [hit    "[h]it"
        stay   "[s]tay"
        double "[d]ouble"
        options (set (list hit stay))
        options (if (and first? (>= balance bet)) (conj options double) options)
        points (cards/hand-points hand)]
    (println "Your hand:" (cards/display-hand hand))
    (cond (> points 21)
          (do (println "Bust.")
              (list balance (rest deck) points 0 false))

          (and first? (= points 21))
          (do (println "Blackjack!")
              ;; Returns:
              (list balance (rest deck) points (* 2.5 bet) false))

          :else
          (loop []
            (println "Please enter an action:" (join ", " options))
            (let [action (lower-case (read-line))]
              (cond (and (= action "h") (contains? options hit))
                    (do (println "You're delt:" (cards/display-card (first deck)))
                        (play-card (cons (first deck) hand) (rest deck) false balance bet))

                    (and (= action "s") (contains? options stay))
                    ;; Returns:
                    (list balance (rest deck) points bet true)

                    (and (= action "d") (contains? options double))
                    (do (let [hand (cons (first deck) hand)
                              points (cards/hand-points hand)]
                          (println "You're delt" (cards/display-card (first deck)))
                          (if (> (cards/hand-points hand) 21)
                            (do (println "Bust.")
                                ;; Returns:
                                (list balance (rest deck) points 0 false)))
                          ;; Returns:
                          (list (- balance bet) (rest deck) points (* 2 bet) true)))

                    :else (recur)))))))

(defn dealer-play-hand
  "Has the dealer play a hand given hand and deck
  Always hits on <soft 17
  Returns point value of hand (0 if bust)"
  [hand deck]
  (println "Dealer's hand:" (cards/display-hand hand))
  (let [points (cards/hand-points hand)]
    (cond (> points 21)
          (do (println "Dealer busts!")
              0)

          (or (< points 17) (and (= points 17) (cards/hand-soft? hand)))
          (do (println "Dealer hits.")
              (println "Dealer draws " (cards/display-card (first deck)))
              (dealer-play-hand (cons (first deck) hand) (rest deck)))

          :else
          (do (println "Dealer stays.")
              points))))

(defn play-hand
  "Walks the player through a hand of blackjack"
  [balance]
  (let [bet (loop []
              (println "Bet: ")
              ;; Note read-string is an ARBITRARY CODE EXECUTION venerability
              (let [parsed-bet (read-string (read-line))]
                (cond (not (number? parsed-bet))
                      (do
                        (println "Sorry, I didn't understand that. Please enter a number.")
                        (recur))

                      (> parsed-bet balance)
                      (do
                        (println "You can't bet more units than you have in your balanace.")
                        (recur))

                      (<= parsed-bet 0)
                      (do
                        (println "You can't bet 0 or less units")
                        (recur))

                      :else parsed-bet
                      )))
        balance (- balance bet)
        [deck player-hand dealer-hand] (deal)
        [balance deck player-points bet dealer?] (play-card player-hand deck true balance bet)]
    (if (not dealer?)
      (+ balance bet)
      (let [dealer-points (dealer-play-hand dealer-hand deck)]
        (cond
          (= dealer-points player-points)
          (do
            (println "Draw.")
            (+ balance bet))

          (< dealer-points player-points)
          (do
            (println "You win!")
            (+ balance (* 2 bet)))
          :else
          (do
            (println "House wins.")
            balance))))))

;; player bets
;; dealt 2 cards face up
;; dealer one down, one up
;; hit: always
;; stay: always
;; split: two of the same card
;; double: showing 9, 10, or 11

;; Dealer: always hits soft 17 or less, stays on hard 17 or more
(defn -main
  "Main game loop, and blackjack game entry point"
  [& _]
  (loop [account-balance 1000]
    (cond (and (<= account-balance 0)
               (y-n "You appear to have run out of money."
                    "\nDo you wish to reset your balance to 1000 units?"))
          (recur 1000)

          (and (> account-balance 0)
               (y-n "Your balance is " account-balance " units. Play a hand?"))
          (recur (play-hand account-balance))

          :else (println "Game over."))))


src/blackjack_clj/cards.clj
(ns blackjack-clj.cards
  (:require [clojure.math.combinatorics :as combo]
            [blackjack-clj.util :refer :all]))

(def hearts "Hearts")
(def spades "Spades")
(def clubs "Clubs")
(def diamonds "Diamonds")

(defn suit
  "Returns the suit of a card"
  [[_ suit]]
  suit)

(defn display-value
  "Returns the value of the card displayed to the player"
  [[value _]]
  (cond
    (= 0 value) "Ace"
    (= 10 value) "Jack"
    (= 11 value) "Queen"
    (= 12 value) "King"
    :else (str (inc value))))

(defn display-card
  "Returns a human-readable version of the card"
  [card]
  (str (display-value card) " of " (suit card)))

(defn card-points
  "The number of points that a card counts for"
  [[value _]]
  (if (= value 0)
    11
    (min (inc value) 10)))

(defn hand-points
  "The number of points that a hand counts for"
  [cards]
  (let [[points num-aces]
        (loop [[card & rest] cards
               num-aces 0
               points-so-far 0]
          (if (nil? card)
            (list points-so-far num-aces)
            (let [points (card-points card)
                  new-sum (+ points points-so-far)]
              (if (= points 11)
                (recur rest (inc num-aces) new-sum)
                (recur rest num-aces new-sum)))))]
    ;; Cut down points if we're over 21 and have aces
    (loop [remain-aces num-aces
           remain-points points]
      (if (or (= remain-aces 0) (<= remain-points 21))
        remain-points
        (recur (dec remain-aces) (- remain-points 10))))))

(defn display-hand
  "Returns a human-readable version of a hand of cards"
  [hand]
  (str (apply str (->> hand
                       (map display-card)
                       (interpose ", ")))
       " ("
       (hand-points hand)
       " points)"))

(defn hand-soft?
  "Returns whether or not the hand contains an ace being
  used as 11 points"
  [hand]
  (let [points-list (map card-points hand)]
    (and
     (in? points-list 11)
     (not= (->> points-list
                (map (fn [points] (if (= points 11) 1 points)))
                (reduce +))
           (hand-points hand)))))

(def unshuffled-deck
  "An unshuffled deck of cards"
  (combo/cartesian-product (range 13) (list hearts spades clubs diamonds)))

(defn shuffled-deck
  "returns a new shuffled deck of cards"
  []
  (shuffle unshuffled-deck))
